app/lib/task/backend.dart

```diff
@@ -432,17 +432,16 @@ class TaskBackend {
       }
 
       // Make changes!
-      state.versions!
-        // Remove versions that have been deselected
-        ..removeWhere((v, _) => deselectedVersions.contains(v))
-        // Add versions we should be tracking
-        ..addAll({
-          for (final v in untrackedVersions)
-            v: PackageVersionStateInfo(
+      // Note: iterating over the tracked versions list keeps the order.
+      state.versions = Map.fromEntries(versions.map((v) {
```

Comment:
Nothing specifies that this property should be sorted.
That's not an invariant, if it's an important invariant, it's an undocumented one.

Given that it's a map, I'd suggest we shouldn't attempt to sort it. Yes, Dart does tend to maintain the ordering of maps, but this is perhaps not something we want to rely on. It's very easy to break this, even if it is documented.

And it's not, see:
https://github.com/dart-lang/pub-dev/blob/bf3d7029ebcdb5ec9ab7bb045f19ed0d6124f792/app/lib/task/models.dart#L106-L108

Reply:
I propose that we sort the versions here:
https://github.com/dart-lang/pub-dev/blob/bf3d7029ebcdb5ec9ab7bb045f19ed0d6124f792/app/lib/task/scheduler.dart#L289C31-L289C46

Or in here:
https://github.com/dart-lang/pub-dev/blob/bf3d7029ebcdb5ec9ab7bb045f19ed0d6124f792/app/lib/task/scheduler.dart#L289C31-L289C46
(in which case we should also document that they are sorted in order of priority).

Should probably use something like [Version.prioritize](https://pub.dev/documentation/pub_semver/latest/pub_semver/Version/prioritize.html).

Or maybe we want to make our own sorting, where we prioritize the latest stable release highest, and then everything else in order of version number.

 * `2.5.0` (this is special, otherwise this is the order of versions)
 * `3.0.0-dev.5`
 * `3.0.0-dev.4`
 * ...
 * `3.0.0-dev.1`
 * `2.4.0`
 * `2.4.0-dev.1`
 * `2.3.0`
 * ...

Or maybe, we prioritize everything in order of [Version.prioritize](https://pub.dev/documentation/pub_semver/latest/pub_semver/Version/prioritize.html) with the exception of the latest pre-release which is prioritized second highest.

 * `2.5.0`
 * `3.0.0-dev.5` (this is special, otherwise this is order of [Version.prioritize](https://pub.dev/documentation/pub_semver/latest/pub_semver/Version/prioritize.html))
 * `2.4.0`
 * ...
 * `2.3.0`
 * `3.0.0-dev.4`
 * ...
 * `3.0.0-dev.1`
 * `2.4.0-dev.1`
 * ...

--------------

Regardless, let's not try to maintain an ordering in datastore. It's much safer to sort all entries when we create the payload we send to the worker.

Yes, I realize we'll have to parse the versions, but this is extremely cheap (compared to actually downloading an analyzing the versions). Also these are sort lists.